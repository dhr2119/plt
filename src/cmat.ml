open Scanner
open Parser
open Ast
open Utils
open Codegen
open Semant
open Llvm

type action = Ast | LLVM_IR | Compile

let _ =
let action = if Array.length Sys.argv > 1 then
  List.assoc Sys.argv.(1)
    [ ("-a",Ast); ("-l",LLVM_IR); ("-c",Compile) ]
  else Compile in
    let lexbuf = Lexing.from_channel stdin in
    let ast = Parser.program Scanner.token lexbuf in

    let gst = Semant.check_var_decls (fst ast) in
      Semant.check_functions gst (snd ast);

    match action with
        Ast -> print_string(Utils.string_of_program ast)
      | LLVM_IR -> print_string(Llvm.string_of_llmodule(Codegen.translate ast))
      | Compile -> let m = Codegen.translate ast in
          Llvm_analysis.assert_valid_module m; (* Useful built-in check *)
          print_module ("hello_world.ll") (m);
