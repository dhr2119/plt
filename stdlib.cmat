matrix int [2,2] addi22(matrix int [2,2] M1, matrix int [2,2] M2) {
    matrix int [2,2] M3;
    int i;
    int j;
    bool b1;
    bool b2;
    b1 = (M1:rows == M2:rows);
    b2 = (M1:cols == M2:cols);
    if(b1 && b2) {
        for(i=0; i < M1:rows; ++i) {
            for(j=0; j < M1:cols; ++j) {
                M3[i,j] = M1[i,j] + M2[i,j];
            }
        }
    }
    return M3;
}

int powi(int i, int n) {
    int k;
    int pow;
    for(k=0; k<n; ++k) {
        pow = pow*k;
    }
    return pow;
}

float powf(float f, int n) {
    int k;
    float pow;
    for(k=0; k<n; ++k) {
        pow = pow*k;
    }
    return pow;
}


void printi22(matrix int [2,2] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi23(matrix int [2,3] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi24(matrix int [2,4] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi25(matrix int [2,5] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi26(matrix int [2,6] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi27(matrix int [2,7] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi28(matrix int [2,8] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi29(matrix int [2,9] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi32(matrix int [3,2] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi33(matrix int [3,3] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi34(matrix int [3,4] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi35(matrix int [3,5] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi36(matrix int [3,6] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi37(matrix int [3,7] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi38(matrix int [3,8] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi39(matrix int [3,9] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi42(matrix int [4,2] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi43(matrix int [4,3] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi44(matrix int [4,4] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi45(matrix int [4,5] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi46(matrix int [4,6] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi47(matrix int [4,7] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi48(matrix int [4,8] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi49(matrix int [4,9] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi52(matrix int [5,2] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi53(matrix int [5,3] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi54(matrix int [5,4] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi55(matrix int [5,5] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi56(matrix int [5,6] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi57(matrix int [5,7] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi58(matrix int [5,8] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi59(matrix int [5,9] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi62(matrix int [6,2] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi63(matrix int [6,3] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi64(matrix int [6,4] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi65(matrix int [6,5] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi66(matrix int [6,6] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi67(matrix int [6,7] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi68(matrix int [6,8] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi69(matrix int [6,9] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi72(matrix int [7,2] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi73(matrix int [7,3] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi74(matrix int [7,4] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi75(matrix int [7,5] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi76(matrix int [7,6] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi77(matrix int [7,7] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi78(matrix int [7,8] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi79(matrix int [7,9] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi82(matrix int [8,2] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi83(matrix int [8,3] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi84(matrix int [8,4] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi85(matrix int [8,5] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi86(matrix int [8,6] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi87(matrix int [8,7] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi88(matrix int [8,8] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi89(matrix int [8,9] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi92(matrix int [9,2] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi93(matrix int [9,3] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi94(matrix int [9,4] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi95(matrix int [9,5] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi96(matrix int [9,6] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi97(matrix int [9,7] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi98(matrix int [9,8] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}

void printi99(matrix int [9,9] M) {
    int i;
    int j;
    for(i=0; i < M:rows; ++i) {
        for(j=0; j < M:cols; ++j) {
            print_int(M[i,j]);
        }
        print_string("");
    }
}
