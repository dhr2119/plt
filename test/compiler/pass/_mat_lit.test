#include <../../stdlib.cmat>;

int main() {
    int i;
    int j;
    matrix int [2,3] mi;
    matrix int [2,3] mj;
    matrix int [2,3] ma;
    matrix int [2,3] ms;
    mi = [1,2,3;4,5,6];
    mj = [1,2,3;4,5,6];
    ma = mi+mj;
    ms = mi-mj;
    print_string("mi:");
    printi23(mi);
    print_string("mj:");
    printi23(mj);
    print_string("ma:");
    printi23(ma);
    print_string("ms:");
    printi23(ms);
    return 0;
}