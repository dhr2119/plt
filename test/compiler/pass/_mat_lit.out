1 2 3 
4 5 6 
1.300000 2.200000 
3.100000 4.600000 
5.200000 6.500000 
